bridge-history-api/internal/orm/cross_message.go

```diff
@@ -0,0 +1,395 @@
+		Columns:   []clause.Column{{Name: "message_hash"}},
+		DoUpdates: clause.AssignmentColumns([]string{"sender", "receiver", "token_type", "l2_block_number", "l2_tx_hash", "l1_token_address", "l2_token_address", "token_ids", "token_amounts", "message_type", "tx_status", "block_timestamp", "message_from", "message_to", "message_value", "message_data", "merkle_proof", "message_nonce"}),
+	})
+	for _, message := range messages {
+		if err := db.Create(message).Error; err != nil {
+			return fmt.Errorf("failed to insert message, message: %+v, error: %w", message, err)
+		}
+	}
+	return nil
+}
+
+// InsertFailedGatewayRouterTxs inserts a list of transactions that failed to interact with the gateway router into the database.
+// These failed transactions are only fetched once, so they are inserted without checking for duplicates.
+// To resolve unique index confliction, a random UUID will be generated and used as the MessageHash.
+func (c *CrossMessage) InsertFailedGatewayRouterTxs(ctx context.Context, messages []*CrossMessage, dbTX ...*gorm.DB) error {
+	db := c.db
+	if len(dbTX) > 0 && dbTX[0] != nil {
+		db = dbTX[0]
+	}
+
+	db = db.WithContext(ctx)
+	db = db.Model(&CrossMessage{})
+	for _, message := range messages {
+		u := uuid.New()
+		message.MessageHash = u.String()
+		if err := db.Create(message).Error; err != nil {
+			return fmt.Errorf("failed to insert failed gateway router txs, message: %+v, error: %w", message, err)
+		}
+	}
+	return nil
+}
+
+// InsertOrUpdateL2RelayedMessagesOfL1Deposits inserts or updates the database with a list of L2 relayed messages related to L1 deposits.
+func (c *CrossMessage) InsertOrUpdateL2RelayedMessagesOfL1Deposits(ctx context.Context, l2RelayedMessages []*CrossMessage, dbTX ...*gorm.DB) error {
+	db := c.db.WithContext(ctx)
+	db = db.Model(&CrossMessage{})
+	db = db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "message_hash"}},
+		DoUpdates: clause.AssignmentColumns([]string{"l2_block_number", "l2_tx_hash", "tx_status"}),
+	})
+	for _, l2RelayedMessage := range l2RelayedMessages {
+		result := db.Create(l2RelayedMessage)
+		if result.Error != nil {
+			return fmt.Errorf("failed to update L2 relayed message of L1 deposit, L2 relayed message: %+v, error: %w", l2RelayedMessage, result.Error)
+		}
+	}
+	return nil
+}
+
+// InsertOrUpdateL1RelayedMessagesOfL2Withdrawals inserts or updates the database with a list of L1 relayed messages related to L2 withdrawals.
+func (c *CrossMessage) InsertOrUpdateL1RelayedMessagesOfL2Withdrawals(ctx context.Context, l1RelayedMessages []*CrossMessage, dbTX ...*gorm.DB) error {
+	db := c.db
+	if len(dbTX) > 0 && dbTX[0] != nil {
+		db = dbTX[0]
+	}
+	db = db.WithContext(ctx)
+	db = db.Model(&CrossMessage{})
+	db = db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "message_hash"}},
+		DoUpdates: clause.AssignmentColumns([]string{"l1_block_number", "l1_tx_hash", "tx_status"}),
+	})
+	for _, l1RelayedMessage := range l1RelayedMessages {
```

Comment:
don't need for..loop create. just `result := db.Create(l1RelayedMessages)`. this logic is not similar with chain-monitor